skdownscale/pointwise_models/quantile.py

```diff
@@ -398,6 +415,175 @@ def _more_tags(self):
+    n_endpoints : int
+        Number of endpoints to include when extrapolating the tails of the cdf. Usused if ``extrapolate`` is None. Default is 10.
+
+    Attributes
+    ----------
+    cdf_ : Cdf
+        NamedTuple representing the fit cdf
+    """
+
+    _fit_attributes = ['cdf_']
+
+    def __init__(self, *, alpha=0.4, beta=0.4, extrapolate='both', n_endpoints=10):
+
+        self.alpha = alpha
+        self.beta = beta
+
+        self.extrapolate = extrapolate
+        self.n_endpoints = n_endpoints
+
+    def fit(self, X, y=None):
+        """Compute CDF and plotting positions for X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, 1)
+            The data used to scale along the features axis. If a sparse
+            matrix is provided, it will be converted into a sparse
+            ``csc_matrix``. Additionally, the sparse matrix needs to be
+            nonnegative if `ignore_implicit_zeros` is False.
+        y : None
+            Ignored.
+
+        Returns
+        -------
+        self : object
+           Fitted transformer.
+        """
+
+        X = check_array(X, ensure_2d=True)
+        if X.shape[1] > 1:
+            raise ValueError('CunnaneTransformer.fit() only supports a single feature')
+        X = X[:, 0]
+
+        self.cdf_ = Cdf(plotting_positions(len(X)), np.sort(X))
+        return self
+
+    def transform(self, X):
+        """Perform the quantile transform.
+
+        Parameters
+        ----------
+        X : array_like, shape [n_samples, 1]
+            Samples.
+
+        Returns
+        -------
+        y : ndarray of shape (n_samples, )
+            Transformed data
+        """
+
+        X = check_array(X, ensure_2d=True)
+
+        if X.shape[1] > 1:
+            raise ValueError('CunnaneTransformer.transform() only supports a single feature')
+        X = X[:, 0]
+
+        left = -np.inf if self.extrapolate in ['min', 'both'] else None
+        right = np.inf if self.extrapolate in ['max', 'both'] else None
+
+        pps = np.interp(X, self.cdf_.vals, self.cdf_.pp, left=left, right=right)
+
+        if np.isinf(pps).any():
+            lower_inds = np.nonzero(-np.inf == pps)[0]
+            upper_inds = np.nonzero(np.inf == pps)[0]
+            model = LinearRegression()
+            if len(lower_inds):
+                s = slice(None, self.n_endpoints)
+                model.fit(self.cdf_.vals[s].reshape(-1, 1), self.cdf_.pp[s].reshape(-1, 1))
+                pps[lower_inds] = model.predict(X[lower_inds].values.reshape(-1, 1)).squeeze()
+            if len(upper_inds):
+                s = slice(-self.n_endpoints, None)
+                model.fit(self.cdf_.vals[s].reshape(-1, 1), self.cdf_.pp[s].reshape(-1, 1))
+                pps[upper_inds] = model.predict(X[upper_inds].values.reshape(-1, 1)).squeeze()
+
+        return pps.reshape(-1, 1)
+
+    def fit_transform(self, X, y=None):
```

Comment:
having this functionality is great. here (and elsewhere) would typing be useful? i'm thinking about if the docstring says that `X` is array-like then maybe we can specify that here